File: go/lazyk.go
package lazyk

import (
	"bufio"
	"io"
	"os"
)

type LazyK struct {
	reader   *bufio.Reader
	writer   *bufio.Writer
	textmode bool
}

func (lk *LazyK) readc() int {
	lk.writer.Flush()
	b, err := lk.reader.ReadByte()
	if err != nil {
		if err == io.EOF {
			return 256
		}
		panic(err)
	}
	if b == 13 && lk.textmode {
		return lk.readc()
	}
	return int(b)
}

func (lk *LazyK) writec(c int) {
	lk.writer.WriteByte(byte(c))
	if c == 10 {
		lk.writer.Flush()
	}
}

func (lk *LazyK) Eval(e *Expr) *Expr {
	cur := e
	var stack []*Expr
	for {
		cur = cur.dropI1()
		for cur.etype == ET_App {
			stack = append(stack, cur)
			cur = cur.arg1.dropI1()
		}
		if len(stack) == 0 {
			return cur
		}
		a := cur
		cur, stack = stack[len(stack)-1], stack[:len(stack)-1]
		cur.arg1 = a
		lk.evalPrimitive(cur)
	}
}

func (lk *LazyK) evalPrimitive(e *Expr) {
	lhs := e.arg1
	rhs := e.arg2.dropI1()
	e.arg1 = nil
	e.arg2 = nil
	switch lhs.etype {
	case ET_False:
		e.etype = ET_I
		return
	case ET_I:
		e.etype = ET_I1
		e.arg1 = rhs
		return
	case ET_K:
		e.etype = ET_K1
		e.arg1 = rhs
		return
	case ET_K1:
		e.etype = ET_I1
		e.arg1 = lhs.arg1
		return
	case ET_S:
		e.etype = ET_S1
		e.arg1 = rhs
		return
	case ET_S1:
		e.etype = ET_S2
		e.arg1 = lhs.arg1
		e.arg2 = rhs
		return
	case ET_S2:
		e.arg1 = lhs.arg1.Apply(rhs)
		e.arg2 = lhs.arg2.Apply(rhs)
		return
	case ET_Iota:
		e.arg1 = rhs.Apply(E_S)
		e.arg2 = E_K
		return
	case ET_Read:
		cn := cnum(lk.readc())
		readnext := expr0(ET_Read)
		e.arg1 = rhs.Apply(cn)
		e.arg2 = readnext
		lhs.etype = ET_S2
		lhs.arg1 = expr2(ET_S2, E_I, expr1(ET_K1, cn))
		lhs.arg2 = expr1(ET_K1, readnext)
		return
	case ET_CNum:
		e.etype = ET_CNum1
		e.arg1 = rhs
		e.intval = lhs.intval
		return
	case ET_CNum1:
		if lhs.arg1 == E_Inc {
			if rhs.etype != ET_Num {
				rhs = lk.Eval(rhs)
			}
			if rhs.etype == ET_Num {
				e.etype = ET_Num
				e.intval = lhs.intval + rhs.intval
				return
			} else {
				panic("invalid output format")
			}
		} else {
			f := lhs.arg1
			x := rhs
			for i := 0; i < lhs.intval; i++ {
				x = f.Apply(x)
			}
			e.etype = x.etype
			e.arg1 = x.arg1
			e.arg2 = x.arg2
			e.intval = x.intval
			return
		}
	case ET_Inc:
		if rhs.etype != ET_Num {
			rhs = lk.Eval(rhs)
		}
		if rhs.etype != ET_Num {
			panic("invalid output format")
		}
		e.etype = ET_Num
		e.intval = rhs.intval + 1
		return
	case ET_Num:
		panic("invalid output format")
	default:
		panic("bug: unexpected state")
	}
}

func (lk *LazyK) cnum2int(cn *Expr) int {
	n := lk.Eval(cn.Apply(E_Inc).Apply(E_Num0))
	if n.etype != ET_Num {
		panic("invalid output format")
	}
	return n.intval
}

func (lk *LazyK) printList(lst *Expr) int {
	car := func(e *Expr) *Expr { return e.Apply(E_True) }
	cdr := func(e *Expr) *Expr { return e.Apply(E_False) }
	for {
		c := lk.cnum2int(car(lst))
		if c >= 256 {
			lk.writer.Flush()
			return c - 256
		}
		lk.writec(c)
		lst = cdr(lst)
	}
}

func (lk *LazyK) run(prog *Expr) int {
	return lk.printList(prog.Apply(expr0(ET_Read)))
}

func (lk *LazyK) RunFile(srcpath string) int {
	fi, err := os.Open(srcpath)
	if err != nil {
		panic(err)
	}
	defer func() {
		if err := fi.Close(); err != nil {
			panic(err)
		}
	}()
	result := lk.run(Parse(fi))
	return result
}

func (lk *LazyK) RunString(src string) int {
	prog := ParseString(src)
	return lk.run(prog)
}

func NewLazyK(rr io.Reader, wr io.Writer, textmode bool) *LazyK {
	brr := bufio.NewReader(rr)
	bwr := bufio.NewWriter(wr)
	return &LazyK{brr, bwr, textmode}
}

File: csharp/lazyk-cs/LazyK.cs
using System;
using System.Buffers;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using static lazyk_cs.Expr;

namespace lazyk_cs
{
    public class LazyK
    {
        private Stream input;
        private byte[] input_buf;
        private int input_len;
        private int input_pos;
        private bool input_isEOF;
        private bool textmode;
        private Stream output;
        private byte[] output_buf;
        private int output_pos;

        public LazyK(Stream input, Stream output, bool textmode)
        {
            if (!input.CanRead)
            {
                throw new Exception(String.Format("{0} is not readable", input.ToString()));
            }
            if (!output.CanWrite)
            {
                throw new Exception(String.Format("{0} is not readable", output.ToString()));
            }
            this.input = input;
            this.input_isEOF = false;
            this.input_buf = new byte[1024];
            this.input_len = 0;
            this.input_pos = 0;
            this.output = output;
            this.output_buf = new byte[1024];
            this.output_pos = 0;
            this.textmode = textmode;
        }

        #region IO
        private int readc()
        {
            if (input_pos == input_len && !input_isEOF)
            {
                flush();
                input_pos = 0;
                input_len = input.Read(input_buf, 0, input_buf.Length);
                if (input_len == 0)
                {
                    input_isEOF = true;
                    return 256;
                }
            }
            if (input_pos < input_len)
            {
                int c = input_buf[input_pos++];
                if (c == 13 && textmode)
                {
                    return readc();
                }
                else
                {
                    return c;
                }
            }
            else
            {
                return 256;
            }
        }

        private void flush()
        {
            output.Write(output_buf, 0, output_pos);
            output_pos = 0;
            output.Flush();
        }

        private void writec(byte b)
        {
            output_buf[output_pos++] = b;
            if (b == 10 || output_pos == output_buf.Length)
            {
                flush();
            }
        }

        public int PrintList(Expr lst)
        {
            while (true)
            {
                int c = ChurchNumToInt(lst.Apply(Expr.K));
                if (c < 256)
                {
                    writec((byte)c);
                }
                else
                {
                    flush();
                    return c - 256;
                }
                lst = Eval(lst.Apply(Expr.False));
            }
        }
        #endregion


        #region
        public Expr Eval(Expr e)
        {
            var stack = new Stack<Expr>();
            Expr cur = e;
            while (true)
            {
                cur = cur.DropI1();
                while (cur.Type == EType.App)
                {
                    Debug.Assert(cur.Arg1 != null && cur.Arg2 != null);
                    stack.Push(cur);
                    cur = cur.Arg1.DropI1();
                }
                if (stack.Count == 0)
                {
                    return cur;
                }
                var a = cur;
                cur = stack.Pop();
                cur.Arg1 = a;
                evalPrimitive(cur);
            }
        }

        private void evalPrimitive(Expr e)
        {
            Debug.Assert(e.Type == EType.App && e.Arg1 != null && e.Arg2 != null);
            Expr lhs = e.Arg1;
            Expr rhs = e.Arg2;
            e.Arg1 = e.Arg2 = null;
            Debug.Assert(lhs.Type != EType.I1);
            switch (lhs.Type)
            {
                case EType.False:
                    e.Type = EType.I;
                    break;
                case EType.I:
                    e.Type = EType.I1;
                    e.Arg1 = rhs;
                    break;
                case EType.K:
                    e.Type = EType.K1;
                    e.Arg1 = rhs;
                    break;
                case EType.K1:
                    e.Type = EType.I1;
                    e.Arg1 = lhs.Arg1;
                    break;
                case EType.S:
                    e.Type = EType.S1;
                    e.Arg1 = rhs;
                    break;
                case EType.S1:
                    Debug.Assert(lhs.Arg1 != null);
                    e.Type = EType.S2;
                    e.Arg1 = lhs.Arg1;
                    e.Arg2 = rhs;
                    break;
                case EType.S2:
                    Debug.Assert(lhs.Arg1 != null && lhs.Arg2 != null);
                    e.Arg1 = lhs.Arg1.Apply(rhs);
                    e.Arg2 = lhs.Arg2.Apply(rhs);
                    break;
                case EType.Iota:
                    e.Arg1 = rhs.Apply(S);
                    e.Arg2 = K;
                    break;
                case EType.Read:
                    Expr ch = CNums[readc()];
                    Expr readnext = new Expr(EType.Read);
                    e.Arg1 = rhs.Apply(ch);
                    e.Arg2 = readnext;
                    lhs.Type = EType.S2;
                    lhs.Arg1 = new Expr(EType.S2, Expr.I, new Expr(EType.K1, ch));
                    lhs.Arg2 = new Expr(EType.K1, readnext);
                    lhs.IntVal = -1;
                    break;
                case EType.CNum:
                    Debug.Assert(lhs.IntVal >= 0);
                    e.Type = EType.CNum1;
                    e.Arg1 = rhs;
                    e.IntVal = lhs.IntVal;
                    break;
                case EType.CNum1:
                    Debug.Assert(lhs.IntVal >= 0 && lhs.Arg1 != null);
                    if (lhs.Arg1.Type == EType.Inc)
                    {
                        if (rhs.Type != EType.Num) rhs = Eval(rhs);
                        if (rhs.Type == EType.Num)
                        {
                            e.Type = EType.Num;
                            e.IntVal = lhs.IntVal + rhs.IntVal;
                        }
                        else
                        {
                            throw new Exception(String.Format("invalid output format (attempted to apply inc to a non-number): ({0})", rhs.ToString()));
                        }
                    }
                    else
                    {
                        var f = lhs.Arg1;
                        var x = rhs;
                        for (int i = 0; i < lhs.IntVal; i++)
                        {
                            x = f.Apply(x);
                        }
                        e.Type = x.Type;
                        e.Arg1 = x.Arg1;
                        e.Arg2 = x.Arg2;
                        e.IntVal = x.IntVal;
                    }
                    break;
                case EType.Inc:
                    if (rhs.Type != EType.Num)
                    {
                        rhs = Eval(rhs);
                    }
                    if (rhs.Type == EType.Num)
                    {
                        e.Type = EType.Num;
                        e.IntVal = rhs.IntVal + 1;
                    }
                    else
                    {
                        throw new Exception(String.Format("invalid output format (attempted to apply inc to a non-number): ({0})", rhs.ToString()));
                    }
                    break;
                case EType.Num:
                    throw new Exception(String.Format("invalid output format ({0} is not applicable)", lhs.ToString()));
                default:
                    throw new Exception(String.Format("unexpected state (lhs={0} rhs={1})", lhs.ToString(), rhs.ToString()));
            }
            return;
        }

        public int ChurchNumToInt(Expr e)
        {
            var n = Eval(e.Apply(Expr.Inc).Apply(Expr.Num0));
            if (n.Type == EType.Num)
            {
                return n.IntVal;
            }
            else
            {
                throw new Exception(String.Format("invalid output format (result was not a number): {0}", e.Dump(-1)));
            }
        }
        #endregion
        public int Run(string src)
        {
            try
            {
                var prog = Parser.Parse(src).Apply(new Expr(Expr.EType.Read));
                int result = PrintList(Eval(prog));
                return result;
            }
            catch (Exception e)
            {
                output.Flush();
                Console.Error.WriteLine(e.ToString());
                return 1;
            }

        }
    }
}
